src/library/scala/collection/immutable/ArraySeq.scala

```diff
@@ -77,12 +77,65 @@ sealed abstract class ArraySeq[+A]
     ArraySeq.unsafeWrapArray(unsafeArray.appended[Any](elem)).asInstanceOf[ArraySeq[B]]
 
   override def appendedAll[B >: A](suffix: collection.IterableOnce[B]): ArraySeq[B] = {
-    val b = ArrayBuilder.make[Any]
-    val k = suffix.knownSize
-    if(k >= 0) b.sizeHint(k + unsafeArray.length)
-    b.addAll(unsafeArray)
-    b.addAll(suffix)
-    ArraySeq.unsafeWrapArray(b.result()).asInstanceOf[ArraySeq[B]]
+    def genericResult = {
+      val b = ArrayBuilder.make[Any]
+      val k = suffix.knownSize
+      if (k == 0)
+        this
+      else {
+        if(k >= 0) b.sizeHint(k + unsafeArray.length)
+        b.addAll(unsafeArray)
+        b.addAll(suffix)
+        ArraySeq.unsafeWrapArray(b.result()).asInstanceOf[ArraySeq[B]]
+      }
+    }
+
+    suffix match {
+
+      case that: ArraySeq[_] =>
+        // Optimise concatenation of two ArraySeqs
+        // For ArraySeqs with sizes of [100, 1000, 10000] this is [3.5, 4.1, 5.2]x as fast
+
+        if (isEmpty)
+          that.asInstanceOf[ArraySeq[B]]
+        else if (that.isEmpty)
+          this
+        else {
+          val thisIsObj = this.unsafeArray.isInstanceOf[Array[AnyRef]]
+          val thatIsObj = that.unsafeArray.isInstanceOf[Array[AnyRef]]
+          val mismatch = thisIsObj != thatIsObj
+          if (mismatch)
+
+            // Combining primatives and objects
+            genericResult
+
+          else if (thisIsObj) {
+
+            // A and B are objects
+            val ax = this.unsafeArray.asInstanceOf[Array[A]]
+            val ay = that.unsafeArray.asInstanceOf[Array[B]]
+            val len = ax.length + ay.length
+            val a = new Array[AnyRef](len)
+            System.arraycopy(ax, 0, a, 0, ax.length)
+            System.arraycopy(ay, 0, a, ax.length, ay.length)
+            ArraySeq.unsafeWrapArray(a).asInstanceOf[ArraySeq[B]]
+
+          } else {
+
+            // A is a primative and B = A. Use this instance's protected ClassTag.
```

Comment:
I can see the value of building a primitive array, but doing it only here seems a bit ad-hoc / inconsistent with the rest of the code. Other operations (`updated`, `appended`, `prepended`, `prependedAll`) also build an object array. For example, `ArraySeq(1,2).updated(0,3).unsafeArray.getClass` is an `Array[Object]`.

So what do you think about keeping primitive arrays for these other operations?

And also, should we factor out some of the code in `appendedAll` and also apply the improvements to `prependedAll`?

Reply:
Absolutely! I'll get to it soon

Reply:
Updated. Please take another look @lrytz. I barely know anything about the binary compatibility intricacies so I'm not sure if having a `private[ArraySeq]` method is going to be a problem for MIMA? If so could you please advise an alternative.

Also regarding `.updated` boxing, I'll submit a new PR for that.